src/main.cpp

```diff
@@ -4678,14 +4678,19 @@ bool AskForOutstandingBlocks(uint256 hashStart)
 }
 
 
-
+void clearOrphanBlocks() {
+    for(auto it = mapOrphanBlocks.begin(); it != mapOrphanBlocks.end(); it++) {
+        delete it->second;
+    }
+    mapOrphanBlocks.clear();
```

Comment:
I think we have to clear `mapOrphanBlocksByPrev` here as well as it seems to hold the same pointer.

Reply:
While we're at it, do a `LOCK(cs_main)` when entering the function.